src/all/namicomi/src/eu/kanade/tachiyomi/extension/all/namicomi/Namicomi.kt

```diff
@@ -0,0 +1,386 @@
+     */
+    private fun paginatedChapterListRequest(mangaId: String, offset: Int): Request {
+        val url = NamicomiConstants.apiChapterUrl.toHttpUrl().newBuilder()
+            .addQueryParameter("titleId", mangaId)
+            .addQueryParameter("includes[]", NamicomiConstants.organization)
+            .addQueryParameter("limit", "500")
+            .addQueryParameter("offset", offset.toString())
+            .addQueryParameter("translatedLanguages[]", extLang)
+            .addQueryParameter("order[volume]", "desc")
+            .addQueryParameter("order[chapter]", "desc")
+            .toString()
+
+        return GET(url, headers, CacheControl.FORCE_NETWORK)
+    }
+
+    /**
+     * Requests information about gated chapters (requiring payment & login).
+     */
+    private fun accessibleChapterListRequest(chapterIds: List<String>): Request {
+        return POST(
+            NamicomiConstants.apiGatingCheckUrl,
+            headers,
+            chapterIds
+                .map { EntityAccessRequestItemDto(it, NamicomiConstants.chapter) }
+                .let { helper.json.encodeToString(EntityAccessRequestDto(it)) }
+                .toRequestBody(),
+            CacheControl.FORCE_NETWORK,
+        )
+    }
+
+    override fun chapterListParse(response: Response): List<SChapter> {
+        if (response.code == 204) {
+            return emptyList()
+        }
+
+        val mangaId = response.request.url.toString()
+            .substringBefore("/chapter")
+            .substringAfter("${NamicomiConstants.apiMangaUrl}/")
+
+        val chapterListResponse = response.parseAs<ChapterListDto>()
+        val chapterListResults = chapterListResponse.data.toMutableList()
+        var offset = chapterListResponse.offset
+        var hasNextPage = chapterListResponse.hasNextPage
+
+        // Max results that can be returned is 500 so need to make more API
+        // calls if the chapter list response has a next page.
+        while (hasNextPage) {
+            offset += chapterListResponse.limit
+
+            val newRequest = paginatedChapterListRequest(mangaId, offset)
+            val newResponse = client.newCall(newRequest).execute()
+            val newChapterList = newResponse.parseAs<ChapterListDto>()
+            chapterListResults.addAll(newChapterList.data)
+
+            hasNextPage = newChapterList.hasNextPage
+        }
+
+        // If there are no chapters, don't attempt to check gating
+        if (chapterListResults.isEmpty()) {
+            return emptyList()
+        }
+
+        val gatingCheckRequest = accessibleChapterListRequest(chapterListResults.map { it.id })
+        val gatingCheckResponse = client.newCall(gatingCheckRequest).execute()
+        val accessibleChapterMap = gatingCheckResponse.parseAs<EntityAccessMapDto>()
+            .data?.attributes?.map ?: emptyMap()
+
+        return chapterListResults.mapNotNull {
+            val isAccessible = accessibleChapterMap.getOrElse(it.id) { true }
```

Comment:
is it correct to assume it is accessible if id isn't in the map?

Reply:
In theory, this should never happen. But I've updated it to consider those chapters as inaccessible

Reply:
then change it to:
```kt
val isAccessible = accessibleChapterMap[it.id]!!
```